### PR TITLE
Fix/msg data extract32 usage

### PR DESCRIPTION
**Pull request by skellet0r**
### What I did

Simply added a length getter and an element getter for calldata extract32 usage. 

Fixes: #2423 

### How I did it

Length is simple, we just use the `calldatasize` op, for getting  an element, we simply use `calldataload`

### How to verify it

I added two additional test cases, where I load  a uint256 and an address.
- Some other corner cases may be requested if needed

### Description for the changelog

- feat: allow extract32 usage with msg.data

### Cute Animal Picture

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

File: vyper/builtin_functions/functions.py [hunk 1]
```diff
@@ -851,6 +855,9 @@ def build_LLL(self, expr, args, kwargs, context):
         elif sub.location == "storage":
             lengetter = LLLnode.from_list(["sload", "_sub"], typ=BaseType("int128"))
             elementgetter = _storage_element_getter
+        elif sub.location == "calldata":
+            lengetter = LLLnode.from_list(["calldatasize"], typ="uint256")
+            elementgetter = _calldata_element_getter
         # TODO: unclosed if/elif clause.  Undefined behavior if `sub.location`
```
Comment (charles-cooper) [inline vyper/builtin_functions/functions.py:861]: While you're in here could you get rid of the unclosed else clause? Throw an exception instead of undefined behavior.

File: vyper/builtin_functions/functions.py [hunk 2]
```diff
@@ -820,6 +820,10 @@ def _storage_element_getter(index):
     return LLLnode.from_list(["sload", ["add", "_sub", ["add", 1, index]]], typ=BaseType("int128"),)
 
 
+def _calldata_element_getter(index):
+    return LLLnode.from_list(["calldataload", ["mul", 32, index]], typ=BaseType("int128"))
```
Comment (charles-cooper) [inline vyper/builtin_functions/functions.py:824]: I think the widest type possible --  BaseType("uint256") -- should be used here as int128 is no longer the default in vyper.

**Review by charles-cooper — CHANGES_REQUESTED** [review]
Looks good, couple nits

File: vyper/builtin_functions/functions.py [hunk 3]
```diff
@@ -851,8 +855,12 @@ def build_LLL(self, expr, args, kwargs, context):
         elif sub.location == "storage":
             lengetter = LLLnode.from_list(["sload", "_sub"], typ=BaseType("int128"))
             elementgetter = _storage_element_getter
-        # TODO: unclosed if/elif clause.  Undefined behavior if `sub.location`
-        # isn't one of `memory`/`storage`
+        elif sub.location == "calldata":
+            lengetter = LLLnode.from_list(["calldatasize"], typ="uint256")
```
Comment (charles-cooper) [inline vyper/builtin_functions/functions.py:859]: Sorry I know I looked at this before but I must have missed this. I don't think this is actually correct, since downstream code assumes that the bytestring has a 32-byte header.

Here's the code generated for `extract32(msg.data, 0)`:
```
                    [calldataload, [mul, 32, [div, [clample, 0, [sub, calldatasize, 32]], 32]]]]],
```

I think the kludge would be to return `(add 32 calldatasize)` for `lengetter`. The actual fix would be to refactor Extract32 so it's less broken. I can chat with you about my ideas for that, if you are interested.

**Review by charles-cooper — CHANGES_REQUESTED** [review]

File: vyper/builtin_functions/functions.py (hunk 3 above)
Comment (skellet0r) [reply vyper/builtin_functions/functions.py]: it took me a moment but it clicked, I've implemented the quick fix for the length getter `(add 32 calldatasize)`

File: tests/parser/syntax/test_msg_data.py [hunk 4]
```diff
@@ -159,3 +182,56 @@ def foo(_value: uint256) -> uint256:
...
+    with pytest.raises(TransactionFailed):
+        contract.foo()
+
+    # however, if we append data this will be fine
+    code_1 = """
+@external
+def bar(_addr: address, _num: uint256) -> uint256:
+    result: Bytes[32] = raw_call(
+        _addr,
+        concat(
```
Comment (charles-cooper) [inline tests/parser/syntax/test_msg_data.py:227]: This is an interesting test but I'm not sure we should rely on this behavior. If vyper were stricter it might reject any calldata that is longer than expected.

**Comment by charles-cooper** [comment]
closing in favor of future `extract<N>` feature
